lnwallet/chainfee/filtermanager.go

```diff
@@ -0,0 +1,279 @@
+package chainfee
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/rpcclient"
+)
+
+const (
+	// fetchFilterInterval is the interval between successive fetches of
+	// our peers' feefilters.
+	fetchFilterInterval = time.Minute * 5
+
+	// minNumFilters is the minimum number of feefilters we need during a
+	// polling interval. If we have fewer than this, we won't consider the
+	// data.
+	minNumFilters = 6
+)
+
+var (
+	// errNoData is an error that's returned if fetchMedianFilter is
+	// called and there is no data available.
+	errNoData = fmt.Errorf("no data available")
+)
+
+// filterManager is responsible for determining an acceptable minimum fee to
+// use based on our peers' feefilter values.
+type filterManager struct {
+	// median stores the median of our outbound peer's feefilter values.
+	median    SatPerKWeight
+	medianMtx sync.RWMutex
+
+	fetchFunc func() ([]SatPerKWeight, error)
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+// newFilterManager constructs a filterManager with a callback that fetches the
+// set of peers' feefilters.
+func newFilterManager(cb func() ([]SatPerKWeight, error)) *filterManager {
+	f := &filterManager{
+		quit: make(chan struct{}),
+	}
+
+	f.fetchFunc = cb
+
+	return f
+}
+
+// Start starts the filterManager.
+func (f *filterManager) Start() {
+	f.wg.Add(1)
+	go f.fetchPeerFilters()
+}
+
+// Stop stops the filterManager.
+func (f *filterManager) Stop() {
+	close(f.quit)
+	f.wg.Wait()
+}
+
+// fetchPeerFilters fetches our peers' feefilter values and adds them to our
+// historical values.
+func (f *filterManager) fetchPeerFilters() {
+	filterTicker := time.NewTicker(fetchFilterInterval)
+	defer filterTicker.Stop()
+
+	for {
+		select {
+		case <-filterTicker.C:
+			filters, err := f.fetchFunc()
+			if err != nil {
+				log.Errorf("Encountered err while fetching "+
+					"fee filters: %v", err)
+				return
+			}
+
+			f.updateMedian(filters)
+
+		case <-f.quit:
+			return
+		}
+	}
+}
+
+// fetchMedianFilter fetches the median feefilter value.
+func (f *filterManager) fetchMedianFilter() (SatPerKWeight, error) {
+	f.medianMtx.RLock()
+	defer f.medianMtx.RUnlock()
+
+	if f.median == 0 {
+		// Return errNoFilters so the caller knows to ignore this
+		// output and continue.
+		return 0, errNoData
+	}
+
+	return f.median, nil
+}
+
+// chooseMinFee takes in the minimum relay fee and compares it against our
+// median filter, returning whichever is highest. This function does not need
+// to acquire the lock.
+func (f *filterManager) chooseMinFee(minRelayFee SatPerKWeight) SatPerKWeight {
+	filterFee, err := f.fetchMedianFilter()
+	if err == errNoData {
+		return minRelayFee
+	}
+
+	if filterFee > minRelayFee {
+		return filterFee
+	}
+
+	return minRelayFee
+}
+
+type bitcoindPeerInfoResp struct {
+	Inbound      bool    `json:"inbound"`
+	MinFeeFilter float64 `json:"minfeefilter"`
```

Comment:
Maybe this is a dumb question but can't we also just put `SatPerKWeight` directly into this structure rather than lifting the float64 into that value later? I'd like to minimize the time it takes for us to get values into their semantically valid/rich types.

Reply:
I don't think the json will parse as `SatPerKWeight` is an `int64` under the hood